File: src/design_patterns/behavioral/memento/python/memento.py
class Memento:
    def __init__(self, state):
        self._state = state

    def get_state(self):
        return self._state


class Originator:
    def __init__(self):
        self._state = None

    def set_state(self, state):
        self._state = state

    def save(self):
        return Memento(self._state)

    def restore(self, memento):
        self._state = memento.get_state()


class Caretaker:
    def __init__(self, originator):
        self._mementos = []
        self._originator = originator

    def save_state(self):
        self._mementos.append(self._originator.save())

    def restore_state(self):
        if not self._mementos:
            return

        memento = self._mementos.pop()
        self._originator.restore(memento)

File: src/design_patterns/behavioral/memento/c#/memento.cs
class Memento
{
    private string state;

    public Memento(string state)
    {
        this.state = state;
    }

    public string GetState()
    {
        return state;
    }
}

class Originator
{
    private string state;

    public void SetState(string state)
    {
        this.state = state;
    }

    public Memento Save()
    {
        return new Memento(state);
    }

    public void Restore(Memento memento)
    {
        state = memento.GetState();
    }

    public string GetState()
    {
        return state;
    }
}

class Caretaker
{
    private Stack<Memento> mementos = new Stack<Memento>();
    private Originator originator;

    public Caretaker(Originator originator)
    {
        this.originator = originator;
    }

    public void SaveState()
    {
        mementos.Push(originator.Save());
    }

    public void RestoreState()
    {
        if (mementos.Count == 0)
            return;

        Memento memento = mementos.Pop();
        originator.Restore(memento);
    }
}
